src/authentication/git-auth-scheme.ts

```diff
@@ -0,0 +1,172 @@
+
+interface Auth0LoginResponse {
+  accessToken: string;
+  expisresIn: number;
+}
+
+export class GitAuthScheme {
+  private readonly _getSigningKey: (arg1: string) => Promise<jwksRsa.Key>;
+  private readonly webAuth: auth0.WebAuth;
+  private readonly userLogin: (options: any) => Promise<Auth0LoginResponse>;
+
+  public constructor(
+    auth0ClientId: string,
+    private readonly auth0Domain: string,
+    private readonly auth0Audience: string,
+    private readonly passwordFactory: (username: string) => string,
+    private readonly logger?: logger.Logger,
+  ) {
+    const webAuth = new auth0.WebAuth({
+      domain: auth0Domain.replace(/^https?:\/\//, ''),
+      clientID: auth0ClientId,
+      responseType: 'token',
+      audience: auth0Audience,
+    });
+    this.webAuth = webAuth;
+    this.userLogin = util.promisify(webAuth.client.login.bind(webAuth.client));
+    const keyClient = jwksRsa({
+      cache: true,
+      rateLimit: true,
+      jwksRequestsPerMinute: 2,
+      jwksUri: `${this.auth0Domain}/.well-known/jwks.json`,
+    });
+    keyClient.getSigningKey = keyClient.getSigningKey.bind(keyClient);
+    this._getSigningKey = util.promisify(keyClient.getSigningKey);
+  }
+
+  public async getSigningKey(decoded: any) {
+    if (!decoded || !decoded.header) {
+      console.log(decoded);
+      throw new Error('Invalid JWT');
+    }
+
+    // Only RS256 is supported.
+    if (decoded.header.alg !== 'RS256') {
+      throw new Error('Only RS256 is supported');
+    }
+
+    const key = await this._getSigningKey(decoded.header.kid);
+
+    if (!key.publicKey && !key.rsaPublicKey) {
+      console.log(key);
+      throw new Error('Invalid signing key');
+    }
+    return key.publicKey! || key.rsaPublicKey!;
+  }
+
+  public getScheme() {
+    return (_server: Hapi.Server, _options: any) => ({
+      authenticate: async (request: Hapi.Request, reply: Hapi.IReply) => {
+        try {
+          const { username, password } = this.parseBasicAuth(request);
+          const { accessToken } = await this.login(username, password);
+          console.log(accessToken);
+          const signingKey = await this.getSigningKey(this.decode(accessToken));
+          const credentials = this.verify(accessToken, signingKey);
+          return reply.continue({ credentials });
+        } catch (_error) {
+          this.logger &&
+            this.logger.debug(`Invalid Git request: ${_error.message}`);
+          const error = _error.isBoom
+            ? _error
+            : Boom.create(_error.statusCode || 401, _error.description);
+          return reply(error);
+        }
+      },
+    });
+  }
+
+  public login(username: string, password: string) {
+    if (username.indexOf('@') >= 0) {
+      return this.userLogin({
+        realm: 'Username-Password-Authentication',
+        username,
+        password,
+        scope: 'openid email',
+      });
+    }
+    return this.clientLogin(username, password);
```

Comment:
Added comments to code. The difference is between signed-up regular users and non-interactive clients.